targets/tinyfpga_bx/base.py

```diff
@@ -0,0 +1,127 @@
+        self.clock_domains.cd_por = ClockDomain()
+        reset_delay = Signal(12, reset=4095)
+        self.comb += [
+            self.cd_por.clk.eq(self.cd_sys.clk),
+            self.cd_sys.rst.eq(reset_delay != 0)
+        ]
+        self.sync.por += \
+            If(reset_delay != 0,
+                reset_delay.eq(reset_delay - 1)
+            )
+        self.specials += AsyncResetSynchronizer(self.cd_por, rst | self.reset)
+
+
+class BaseSoC(SoCCore):
+    csr_peripherals = (
+        "spiflash",
+        "info",
+    )
+    csr_map_update(SoCCore.csr_map, csr_peripherals)
+
+    mem_map = {
+        "spiflash": 0x20000000,  # (default shadow @0xa0000000)
+    }
+    mem_map.update(SoCCore.mem_map)
+
+    def __init__(self, platform, **kwargs):
+        if 'integrated_rom_size' not in kwargs:
+            kwargs['integrated_rom_size']=0
+        if 'integrated_sram_size' not in kwargs:
+            kwargs['integrated_sram_size']=0x2800
+
+        # FIXME: Force either lite or minimal variants of CPUs; full is too big.
+
+        platform.add_extension(serial)
+        platform.add_extension(reset)
+        clk_freq = int(16e6)
+
+        # Extra 0x28000 is due to bootloader bitstream.
+        kwargs['cpu_reset_address']=self.mem_map["spiflash"]+platform.gateware_size+0x28000
+        SoCCore.__init__(self, platform, clk_freq, **kwargs)
+
+        self.submodules.crg = _CRG(platform)
+        self.platform.add_period_constraint(self.crg.cd_sys.clk, 1e9/clk_freq)
+
+        # SPI flash peripheral
+        self.submodules.spiflash = spi_flash.SpiFlashSingle(
+            platform.request("spiflash"),
+            dummy=platform.spiflash_read_dummy_bits,
+            div=platform.spiflash_clock_div)
+        self.add_constant("SPIFLASH_PAGE_SIZE", platform.spiflash_page_size)
+        self.add_constant("SPIFLASH_SECTOR_SIZE", platform.spiflash_sector_size)
+        self.register_mem("spiflash", self.mem_map["spiflash"],
+            self.spiflash.bus, size=platform.spiflash_total_size)
+
+        bios_size = 0x8000
+        self.add_constant("ROM_DISABLE", 1)
+        self.add_memory_region("rom", kwargs['cpu_reset_address'], bios_size)
+        self.flash_boot_address = self.mem_map["spiflash"]+platform.gateware_size+bios_size+0x28000
+
+        # We don't have a DRAM, so use the remaining SPI flash for user
+        # program.
+        self.add_memory_region("user_flash",
+            self.flash_boot_address,
+            # Leave a grace area- possible one-by-off bug in add_memory_region?
+            # Possible fix: addr < origin + length - 1
+            platform.spiflash_total_size - (self.flash_boot_address - self.mem_map["spiflash"]) - 0x100)
+
+        # Disable USB activity until we switch to a USB UART.
+        self.comb += [platform.request("usb").pullup.eq(0)]
+
+
+        # Annotate build scripts with required options.
+        # Seed 2 in general works better than seed 1.
+        platform.toolchain.build_template[1] = \
```

Comment:
These should most certainly be part of litex / migen in some way in the future?

Reply:
Based on IRC comments:
```
(6:59:33 PM) mithro: cr1901_modern: I mean there should just be switches which modify the template in some way rather than having to override whole template 
```

It was a conscious design [decision](https://github.com/m-labs/migen/pull/84#discussion_r149899202) to move away from what you're suggesting in Migen. I think you should talk to @jordens and see what he thinks; the `icestorm` backend at present _is_ used to test ideas. I don't know what such a feature you're suggesting would look like short of an explosion of new arguments to the `plat.build` method.

Reply:
Okay, lets worry about that later.